File: ruby/rbfind/lib/rbfind/fileresultformatter.rb
# frozen_string_literal: true

################################################################################
#
# fileresultformatter.rb
#
# Formatter for file results
#
################################################################################

require 'pathname'

require_relative 'color'

module RbFind
  # FileResultFormatter - formats file result instances
  class FileResultFormatter

    attr_reader :settings

    def initialize(settings)
      @settings = settings
      if settings.colorize
        if settings.in_dir_patterns
          # override format_dir_path method in instance with colorizing method
          define_singleton_method(:format_dir_path) { |dir_path| format_dir_path_with_color(dir_path) }
        end
        if settings.in_extensions || settings.in_file_patterns
          # override format_file_name method in instance with colorizing method
          define_singleton_method(:format_file_name) { |file_name| format_file_name_with_color(file_name) }
        end
      end
    end

    def colorize(str, match_start_index, match_end_index)
      prefix = ''
      if match_start_index > 0
        prefix = str[0..(match_start_index - 1)]
      end
      colorized = str[match_start_index..(match_end_index - 1)].green
      suffix = ''
      if match_end_index < str.length
        suffix = str[match_end_index..str.length]
      end
      prefix + colorized + suffix
    end

    def format_dir_path(dir_path)
      dir_path.to_s
    end

    def format_file_name(file_name)
      file_name.to_s
    end

    def format_path(path)
      parent = '.'
      unless path.parent.nil?
        parent = format_dir_path(path.parent)
      end
      file_name = format_file_name(path.basename)
      Pathname(parent).join(file_name).to_s
    end

    def format_file_result(file_result)
      format_path(file_result.path)
    end

    private

    def format_dir_path_with_color(dir_path)
      formatted_dir = dir_path.to_s
      @settings.in_dir_patterns.each do |p|
        m = p.match(formatted_dir)
        if m
          formatted_dir = colorize(formatted_dir, m.begin(0), m.end(0))
          break
        end
      end
      formatted_dir
    end

    def format_file_name_with_color(file_name)
      formatted_file_name = file_name.to_s
      @settings.in_file_patterns.each do |p|
        m = p.match(formatted_file_name)
        if m
          formatted_file_name = colorize(formatted_file_name, m.begin(0), m.end(0))
          break
        end
      end
      unless @settings.in_extensions.empty? or file_name.extname.empty?
        idx = formatted_file_name.rindex('.')
        if idx > 0 && idx < formatted_file_name.length - 1
          formatted_file_name = colorize(formatted_file_name, idx + 1, formatted_file_name.length)
        end
      end
      formatted_file_name
    end

  end
end

File: javascript/jsfind/src/fileresultformatter.js
/*
 * fileresultformatter.js
 *
 * FileResultFormatter class provides formatting of search result instances
 */
const {colorToConsoleColor} = require('./color');
const {ConsoleColor} = require('./consolecolor');
const path = require('path');

class FileResultFormatter {
  'use strict'

  constructor(settings) {
    this.settings = settings;
    if (settings.colorize) {
      if (settings.inDirPatterns.length > 0) {
        this.formatDirPath = function(path) {
          return this.formatDirPathWithColor(path);
        };
      }
      if (settings.inExtensions.length > 0 || settings.inFilePatterns.length > 0) {
        this.formatFileName = function(fileName) {
          return this.formatFileNameWithColor(fileName);
        };
      }
    }
  }

  colorize(s, matchStartIndex, matchEndIndex, color) {
    let prefix = '';
    if (matchStartIndex > 0) {
      prefix = s.slice(0, matchStartIndex);
    }
    let suffix = '';
    if (matchEndIndex < s.length) {
      suffix = s.slice(matchEndIndex);
    }
    return prefix +
      colorToConsoleColor(color) +
      s.slice(matchStartIndex, matchEndIndex) +
      ConsoleColor.RESET +
      suffix;
  }

  formatDirPathWithColor(dirPath) {
    let formattedPath = '.';
    if (dirPath) {
      formattedPath = dirPath;
      for (let p of this.settings.inDirPatterns) {
        let m = p.exec(formattedPath);
        if (m) {
          formattedPath = this.colorize(formattedPath, m.index, m.index + m[0].length, this.settings.dirColor);
          break;
        }
      }
    }
    return formattedPath;
  }

  formatDirPath(dirPath) {
    return dirPath;
  }

  formatFileNameWithColor(fileName) {
    let formattedFileName = fileName;
    for (let p of this.settings.inFilePatterns) {
      let m = p.exec(formattedFileName);
      if (m) {
        formattedFileName = this.colorize(formattedFileName, m.index, m.index + m[0].length, this.settings.fileColor);
        break;
      }
    }
    if (this.settings.inExtensions.length > 0) {
      let idx = formattedFileName.lastIndexOf('.');
      if (idx > 0 && idx < formattedFileName.length - 1) {
        formattedFileName = this.colorize(formattedFileName, idx + 1, formattedFileName.length, this.settings.extColor);
      }
    }
    return formattedFileName;
  }

  formatFileName(fileName) {
    return fileName;
  }

  formatFileResult(result) {
    let parent = this.formatDirPath(result.path);
    let fileName = this.formatFileName(result.fileName);
    return path.join(parent, fileName);
  }
}

exports.FileResultFormatter = FileResultFormatter;
